tensorboard/util/test_case.py

```diff
@@ -0,0 +1,113 @@
+# limitations under the License.
+
+"""TensorBoard test case helper module.
+
+This module provides a TensorBoard base test class with some of the
+niceties of tf.test.TestCase, while only requiring standard unittest
+be installed.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import atexit
+import os
+import shutil
+import tempfile
+import unittest
+
+from tensorboard.util import tb_logging
+
+
+logger = tb_logging.get_logger()
+
+_temp_dir = None
+
+
+def get_temp_dir():
+  """Return a temporary directory for tests to use."""
+  global _temp_dir
+  if not _temp_dir:
+    if os.environ.get('TEST_TMPDIR'):
+      temp_dir = tempfile.mkdtemp(prefix=os.environ['TEST_TMPDIR'])
+    else:
+      temp_dir = tempfile.mkdtemp()
+
+    def delete_temp_dir(dirname=temp_dir):
+      try:
+        shutil.rmtree(dirname)
+      except OSError as e:
+        logger.error('Error removing %s: %s', dirname, e)
+
+    atexit.register(delete_temp_dir)
+    _temp_dir = temp_dir
+
+  return _temp_dir
+
+
+class TestCase(unittest.TestCase):
+  """TensorBoard base test class.
+  This class can lazily create a temporary directory for tests to use.
+  """
+
+  def __init__(self, method='runTest'):
+    super(TestCase, self).__init__(method)
+    self._method = method
+    self._tempdir = None
+
+  def setUp(self):
+    super(TestCase, self).setUp()
+
+  def tearDown(self):
+    super(TestCase, self).tearDown()
+
+  def assertItemsEqual(self, actual, expected, msg=None):
+    """Test that sequence first contains the same elements as second,
+    regardless of their order.
+
+    Same as assertCountEqual in Python 3 with unittest.TestCase."""
+    if hasattr(self, 'assertCountEqual'):
+        self.assertCountEqual(actual, expected, msg)
+    else:
+        super(TestCase, self).assertItemsEqual(actual, expected, msg)
+
+  def assertStartsWith(self, actual, expected_start, msg=None):
+    """Test that string first starts with string second."""
+    if not actual.startswith(expected_start):
+      if msg is None:
+        msg = '{} does not start with {}'.format(actual, expected_start)
+      raise AssertionError(msg)
+
+  def get_temp_dir(self):
+    """Returns a unique temporary directory for the test to use.
+    If you call this method multiple times during in a test, it will return the
+    same folder. However, across different runs the directories will be
+    different. This will ensure that across different runs tests will not be
+    able to pollute each others environment.
+    If you need multiple unique directories within a single test, you should
+    use tempfile.mkdtemp as follows:
+      tempfile.mkdtemp(dir=self.get_temp_dir()):
+    Returns:
+      string, the path to the unique temporary directory created for this test.
+    """
+    if not self._tempdir:
+      self._tempdir = tempfile.mkdtemp(dir=get_temp_dir())
+    return self._tempdir
+
+
+def main(*args, **kwargs):
```

Comment:
Can we remove this function and just call `unittest.main()`?

imho, it seems that a test _case_ module should not need to provide a
test _running_ mechanism, and I don’t see what value this alias
provides.

Reply:
It's primarily that callers don't don't need to also import `unittest`, but  maybe that's not enough benefit. It also matched `tf.test` semantics. Thoughts?

Reply:
Thinking about this a bit more: The existence of `tf.test.main` makes
sense to me because the `tf.test` module _replaces_ the `unittest`
module. In a sense, this new `test_case` module does the same thing, but
its name belies that. Perhaps, then, the right thing to do is to rename
this module to `tensorboard.test`. It doesn’t really provide “utility”
functions in the usual sense, just as TensorFlow’s `tf.test` isn’t
called `tf.test_util`.

What do you think?

(I don’t feel too strongly about this either way, so if you have a
strong preference then I’d probably be happy to accommodate it.)

Reply:
I think you're right. Owning main will let us update / override that as needed to match the `tf.test` semantics. I'll change this to `tensorboard.test` - makes sense to me.